### PR TITLE
Check for incompatible HWLOC

### PR DESCRIPTION
**Pull request by bvanessen**
Added code to check for and throw an error message when an incompatible HWLOC version is loaded.

### CODE REVIEW

File: src/progress.cpp [hunk 1]
```diff
@@ -42,6 +42,30 @@ namespace internal {
...
+#else
+  /* headers are pre-2.0 */
+  if (hwloc_get_api_version() >= 0x20000) {
+    throw_al_exception("HWLOC runtime library "
+                       + print_hwloc_version(hwloc_get_api_version())
+                       + " is more recent than 2.0 but Aluminum was compile with HWLOC API version "
+                       + print_hwloc_version(HWLOC_API_VERSION));
+  }
+#endif
+}
```
Comment (ndryden) [inline src/progress.cpp:67]:
```suggestion
// Adapted from
// https://github.com/open-mpi/hwloc/blob/master/utils/hwloc/misc.h
void check_hwloc_api_version() {
  if ((hwloc_get_api_version() >> 16) != (HWLOC_API_VERSION >> 16)) {
    throw_al_exception("HWLOC runtime library version "
                       + print_hwloc_version(hwloc_get_api_version())
                       + " does not match the version Aluminum was compiled with "
                       + print_hwloc_version(HWLOC_API_VERSION));
  }
}
```

I like this change, but I think this simplifies the code and will be a bit more robust to future HWLOC API version changes.

**Review by ndryden — CHANGES_REQUESTED** [review]
Good catch! This works, but I think my suggestion simplifies things.

File: src/progress.cpp (hunk 1 above)
Comment (bvanessen) [reply src/progress.cpp:67]: That looks good to me.

**Review by ndryden — APPROVED** [review]

File: src/progress.cpp [hunk 2]
```diff
@@ -42,6 +42,21 @@ namespace internal {
 
 namespace {
 
+std::string print_hwloc_version(unsigned long ver) {
+  return std::string{} + std::to_string(ver >> 16) + "." + std::to_string((ver & 0x00ff00) >> 8);
```
Comment (benson31) [inline src/progress.cpp:46]:
```suggestion
  return std::to_string(ver >> 16) + "." + std::to_string((ver & 0x00ff00) >> 8);
```

**Review by benson31 — APPROVED** [review]
LGTM

**Pull request merged**